File: src/test/java/spring/webmvc/application/service/ItemServiceTest.java
package spring.webmvc.application.service;

import java.util.List;
import java.util.Optional;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.util.Pair;

import spring.webmvc.domain.model.entity.Item;
import spring.webmvc.domain.model.enums.Category;
import spring.webmvc.domain.repository.ItemRepository;
import spring.webmvc.presentation.dto.request.ItemSaveRequest;
import spring.webmvc.presentation.dto.response.ItemResponse;
import spring.webmvc.presentation.exception.EntityNotFoundException;

@ExtendWith(MockitoExtension.class)
class ItemServiceTest {

	@InjectMocks
	private ItemService itemService;

	@Mock
	private ItemRepository itemRepository;

	@Test
	@DisplayName("saveItem 은 한개를 저장한다")
	void saveItem_case1() {
		// Given
		ItemSaveRequest request = new ItemSaveRequest("item", "description", 1000, 10, Category.ROLE_BOOK);
		Item item = Item.create("item", "description", 1000, 10, Category.ROLE_BOOK);

		Mockito.when(itemRepository.save(Mockito.any(Item.class))).thenReturn(item);

		// When
		ItemResponse response = itemService.saveItem(request);

		// Then
		Assertions.assertThat(request.name()).isEqualTo(response.name());
		Assertions.assertThat(request.description()).isEqualTo(response.description());
		Assertions.assertThat(request.price()).isEqualTo(response.price());
		Assertions.assertThat(request.quantity()).isEqualTo(response.quantity());
	}

	@Test
	@DisplayName("findItems 은 여러개를 조회한다")
	void findItems_case1() {
		// Given
		Pageable pageable = PageRequest.of(0, 10);
		String name = "Item";
		List<Item> items = List.of(
			Item.create("item1", "description", 1000, 10, Category.ROLE_BOOK),
			Item.create("item2", "description", 2000, 20, Category.ROLE_TICKET),
			Item.create("item3", "description", 3000, 30, Category.ROLE_BOOK)
		);
		Page<Item> page = new PageImpl<>(items, pageable, items.size());

		Mockito.when(itemRepository.findAll(pageable, name)).thenReturn(page);

		// When
		Page<ItemResponse> response = itemService.findItems(pageable, name);

		// Then
		Assertions.assertThat(response.getTotalElements()).isEqualTo(items.size());
	}

	@Test
	@DisplayName("findItem 은 데이터가 있을경우 조회한다")
	void findItem_case1() {
		// Given
		Long itemId = 1L;
		Item item = Item.create("item", "description", 1000, 10, Category.ROLE_BOOK);

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.of(item));

		// When
		ItemResponse response = itemService.findItem(itemId);

		// Then
		Assertions.assertThat(item.getId()).isEqualTo(response.id());
		Assertions.assertThat(item.getName()).isEqualTo(response.name());
		Assertions.assertThat(item.getDescription()).isEqualTo(response.description());
		Assertions.assertThat(item.getPrice()).isEqualTo(response.price());
		Assertions.assertThat(item.getQuantity()).isEqualTo(response.quantity());
	}

	@Test
	@DisplayName("findItem 은 데이터가 없을경우 EntityNotFoundException 던진다")
	void findItem_case2() {
		// Given
		Long itemId = 1L;

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.empty());

		// When & Then
		Assertions.assertThatThrownBy(() -> itemService.findItem(itemId)).isInstanceOf(EntityNotFoundException.class);
	}

	@Test
	@DisplayName("putItem 은 데이터가 없을 경우 생성한다")
	void putItem_case1() {
		// Given
		Long itemId = 1L;
		ItemSaveRequest request = new ItemSaveRequest("NewItem", "description", 2000, 20, Category.ROLE_BOOK);
		Item newItem = Item.create(
			request.name(),
			request.description(),
			request.price(),
			request.quantity(),
			request.category()
		);

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.empty());
		Mockito.when(itemRepository.save(Mockito.any(Item.class))).thenReturn(newItem);

		// When
		Pair<Boolean, ItemResponse> pair = itemService.putItem(itemId, request);
		Boolean isNew = pair.getFirst();
		ItemResponse response = pair.getSecond();

		// Then
		Assertions.assertThat(isNew).isTrue();
		Assertions.assertThat(request.name()).isEqualTo(response.name());
		Assertions.assertThat(request.description()).isEqualTo(response.description());
		Assertions.assertThat(request.price()).isEqualTo(response.price());
		Assertions.assertThat(request.quantity()).isEqualTo(response.quantity());

		Mockito.verify(itemRepository, Mockito.times(1)).save(Mockito.any());
	}

	@Test
	@DisplayName("putItem 은 데이터가 있을 경우 수정한다")
	void putItem_case2() {
		// Given
		Long itemId = 1L;
		Item oldItem = Item.create("oldItem", "description", 1000, 10, Category.ROLE_BOOK);
		ItemSaveRequest request = new ItemSaveRequest("newItem", "description", 2000, 20, Category.ROLE_BOOK);

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.of(oldItem));

		// When
		Pair<Boolean, ItemResponse> pair = itemService.putItem(itemId, request);
		Boolean isNew = pair.getFirst();
		ItemResponse response = pair.getSecond();

		// Then
		Assertions.assertThat(isNew).isFalse();
		Assertions.assertThat(request.name()).isEqualTo(response.name());
		Assertions.assertThat(request.description()).isEqualTo(response.description());
		Assertions.assertThat(request.price()).isEqualTo(response.price());
		Assertions.assertThat(request.quantity()).isEqualTo(response.quantity());

		Mockito.verify(itemRepository, Mockito.never()).save(Mockito.any());
	}

	@Test
	@DisplayName("deleteItem 은 데이터가 없을 경우 EntityNotFoundException 던진다")
	void delete_Item_case1() {
		// Given
		Long itemId = 1L;

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.empty());

		// When & Then
		Assertions.assertThatThrownBy(() -> itemService.deleteItem(itemId)).isInstanceOf(EntityNotFoundException.class);
	}

	@Test
	@DisplayName("deleteItem 은 데이터가 있을 경우 삭제한다")
	void delete_Item_case2() {
		// Given
		Long itemId = 1L;
		Item item = Item.create("item", "description", 1000, 10, Category.ROLE_BOOK);

		Mockito.when(itemRepository.findById(itemId)).thenReturn(Optional.of(item));

		// When
		itemService.deleteItem(itemId);

		// Then
		Mockito.verify(itemRepository, Mockito.times(1)).delete(item);
	}
}

File: spring-web-kotlin/src/test/kotlin/spring/web/kotlin/application/service/ItemServiceTest.kt
package spring.web.kotlin.application.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.spyk
import io.mockk.verify
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import spring.web.kotlin.domain.model.entity.Item
import spring.web.kotlin.domain.model.enums.Category
import spring.web.kotlin.domain.repository.ItemRepository
import spring.web.kotlin.presentation.dto.request.ItemSaveRequest
import spring.web.kotlin.presentation.exception.EntityNotFoundException

class ItemServiceTest : DescribeSpec({
    val itemRepository = mockk<ItemRepository>()
    val itemService = ItemService(itemRepository)

    describe("saveItem 은") {
        it("한개를 저장한다") {
            val request = ItemSaveRequest(
                name = "item1",
                description = "description",
                price = 1000,
                quantity = 10,
                category = Category.ROLE_TICKET
            )

            val item = spyk(
                Item.create(
                    name = request.name,
                    description = request.description,
                    price = request.price,
                    quantity = request.quantity,
                    category = request.category
                )
            ).apply { every { id } returns 1L }

            every { itemRepository.save(any()) } returns item

            itemService.saveItem(request).apply {
                name shouldBe request.name
                description shouldBe request.description
                price shouldBe request.price
                quantity shouldBe request.quantity
            }
        }
    }

    describe("findItems 은") {
        it("여러개를 조회한다") {
            val pageable = PageRequest.of(0, 10)
            val requestName = "Item"
            val items = listOf(
                spyk(
                    Item.create(
                        name = "item1",
                        description = "description1",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns 1L },
                spyk(
                    Item.create(
                        name = "item2",
                        description = "description2",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns 2L },
                spyk(
                    Item.create(
                        name = "item3",
                        description = "description3",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns 3L }
            )
            val page = PageImpl(items, pageable, items.size.toLong())

            every { itemRepository.findAll(pageable, requestName) } returns page

            itemService.findItems(pageable, requestName).apply {
                totalElements shouldBe items.size
                map { it.id }.toList() shouldBe listOf(1L, 2L, 3L)
            }
        }
    }

    describe("findItem 은") {
        val itemId = 5L

        context("데이터가 없을 경우") {
            it("EntityNotFoundException 던진다") {
                every { itemRepository.findByIdOrNull(itemId) } returns null

                shouldThrow<EntityNotFoundException> { itemService.findItem(itemId) }.apply {
                    httpStatus shouldBe HttpStatus.NOT_FOUND
                }
            }
        }

        context("데이터가 있을 경우") {
            it("조회한다") {
                val item = spyk(
                    Item.create(
                        name = "item1",
                        description = "description",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns itemId }

                every { itemRepository.findByIdOrNull(itemId) } returns item

                itemService.findItem(itemId).apply {
                    id shouldBe itemId
                }
            }
        }
    }

    describe("putItem 은") {
        val itemId = 5L
        val request = ItemSaveRequest(
            name = "item2",
            description = "description2",
            price = 10000,
            quantity = 100,
            category = Category.ROLE_TICKET
        )

        context("데이터가 없을 경우") {
            it("생성한다") {
                val item = spyk(
                    Item.create(
                        name = request.name,
                        description = request.description,
                        price = request.price,
                        quantity = request.quantity,
                        category = request.category
                    )
                ).apply { every { id } returns 1L }

                every { itemRepository.findByIdOrNull(itemId) } returns null
                every { itemRepository.save(any()) } returns item

                itemService.putItem(itemId, request).apply {
                    first shouldBe true
                    second.name shouldBe request.name
                    second.description shouldBe request.description
                    second.price shouldBe request.price
                    second.quantity shouldBe request.quantity
                }
            }
        }

        context("데이터가 있을 경우") {
            it("수정한다") {
                val item = spyk(
                    Item.create(
                        name = "item1",
                        description = "description",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns itemId }

                every { itemRepository.findByIdOrNull(itemId) } returns item

                itemService.putItem(itemId, request).apply {
                    first shouldBe false
                    second.name shouldBe request.name
                    second.description shouldBe request.description
                    second.price shouldBe request.price
                    second.quantity shouldBe request.quantity
                }
            }
        }
    }

    describe("deleteItem 은") {
        val itemId = 5L

        context("데이터가 없을 경우") {
            it("EntityNotFoundException 던진다") {
                every { itemRepository.findByIdOrNull(itemId) } returns null

                shouldThrow<EntityNotFoundException> { itemService.deleteItem(itemId) }.apply {
                    httpStatus shouldBe HttpStatus.NOT_FOUND
                }
            }
        }

        context("데이터가 있을 경우") {
            it("삭제한다") {
                val item = spyk(
                    Item.create(
                        name = "item1",
                        description = "description",
                        price = 1000,
                        quantity = 10,
                        category = Category.ROLE_BOOK
                    )
                ).apply { every { id } returns itemId }

                every { itemRepository.findByIdOrNull(itemId) } returns item
                every { itemRepository.delete(item) } returns Unit

                itemService.deleteItem(itemId)

                verify(exactly = 1) { itemRepository.delete(item) }
            }
        }
    }
})
